c/shared/source/uforead/uforead.c

```diff
@@ -1738,22 +1645,110 @@ static int parseXMLFile(ufoCtx h, char* filename, const char* filetype){
 
     if ( cur == NULL ) {
         xmlFreeDoc(doc);
-        fatal(h, ufoErrSrcStream, "Error parsing XML file %s.\n", filename);
+        return NULL;
     }
 
-    if ((!xmlStrEqual((cur)->name, (const xmlChar *) "dict"))) {
-        xmlFreeDoc(doc);
-        fatal(h, ufoErrSrcStream, "Error reading outermost <dict> in %s.\n", filename);
+    if (!strcmp("plist", filetype)) {
+        if ((!xmlStrEqual((cur)->name, (const xmlChar *) "dict"))) {
+            xmlFreeDoc(doc);
+            fatal(h, ufoErrSrcStream, "Error reading outermost <dict> in %s.\n", filename);
+        }
+    }
+    return cur;
+}
+
+static bool strEqual(char* string1, char* string2) {
+    if ((string1 == NULL) || (string2 == NULL)) {
+        return false;
+    }
+    return !(strcmp(string1, string2));
+}
+
+static bool xmlKeyEqual(xmlNodePtr cur, char* name){
+    if (cur != NULL)
+        return xmlStrEqual(cur->name, (const xmlChar *) name);
+    else
+        return false;
+}
+
+static bool xmlAttrEqual(xmlAttr *attr, char* name){
+    if (attr != NULL)
+        return xmlStrEqual(attr->name, (const xmlChar *) name);
+    else
+        return false;
+}
+
+static char* getXmlAttrValue(xmlAttr *attr){
+    return (char*) attr->children->content;
+}
+
+static void parseXMLLib(ufoCtx h, xmlNodePtr cur) {
+    char* keyValue;
+
+    cur = cur->xmlChildrenNode;
+    if (xmlKeyEqual(cur, "dict") && xmlKeyEqual(cur->children, "key")) {
+        cur = cur->children;
+        while (cur != NULL) {
+            char* keyName = parseXMLKeyName(h, cur);
+            cur = cur->next;
+            if (h->parseState.UFOFile == preParsingGLIF) {     /* when called from preParseGLIF */
+                if (strEqual(keyName, "com.adobe.type.cid.CID")) {
+                    keyValue = parseXMLKeyValue(h, cur);
+                    h->parseState.GLIFInfo.currentCID = atoi(keyValue);
```

Comment:
Good catch! I just created a test case for this and it crashes. Would you suggest I add the NULL check to right before setting h->parseState.GLIFInfo.currentCID, or should I return a value like -1 instead of NULL in parseXMLKeyValue so that it fixes this scenario for other callers?

I also realized that if the value of com.adobe.type.cid.CID is a blank value (like the example below) then com.adobe.type.cid.CID is set to 0. I feel like it also probably needs to remain -1 (which is currently what it starts with when unset). 

```
      <key>com.adobe.type.cid.CID</key>
      <integer></integer>
```


I'll look for more places where NULL may be returned from parseXMLKeyValue and the other functions. Thanks!

Reply:
I think being more explicit with NULL checks is probably better for this. Returning "fake" values can eventually result in those values being used.

Reply:
(Although I'd have to look in more detail to give you a conclusive answer ...)